Event Hubs Processor - Add Sample Using the EventProcessorClient with an ASP.NET hosted service

davetrainer opened the pull request:
Addresses [26299](https://github.com/Azure/azure-sdk-for-net/issues/26299)

An example project which demonstrates how to use the EventProcessorClient with an ASP.NET hosted service to accomplish background task processing.

Commit:
Add EventProcessorClient hosted service sample

Commit:
Merge branch 'main' into sample-aspnet-eventprocessorclient

davetrainer commented:
@microsoft-github-policy-service agree

Commit:
fix compliance error and IsPackable

Commit:
add IsPackable = true

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/Azure.Messaging.EventHubs.Samples.Processor.HostedService.csproj:
```diff
@@ -0,0 +1,15 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
+
+  <PropertyGroup>
+    <RequiredTargetFrameworks>net7.0;net6.0</RequiredTargetFrameworks>
```
```suggestion
    <RequiredTargetFrameworks>net8.0;net6.0</RequiredTargetFrameworks>
```
We may as well switch to `net8.0`, as it's now GA.

davetrainer replied:
@jsquire I've left this as net7.0 as I had issues with the SDK. I believe globals.json needs updating first? https://github.com/Azure/azure-sdk-for-net/blob/main/global.json

jsquire replied:
That's my bad.  I'm getting ahead of myself and forgetting that we haven't updated the SDK version yet.  Leaving it as-is makes sense.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/Azure.Messaging.EventHubs.Samples.Processor.HostedService.csproj:
If you'd be so kind, let's shift to a folder structure that allows for the sample and its associated tests to be included here.  For example, the [KeyVault proxy sample](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/keyvault/samples/keyvaultproxy) uses the structure that I'm thinking of - though we'll want to keep the README at the root with its current content.

```text
sdk/eventhub/Azure.Messaging.EventHubs.Processor
|
| -- samples
        | -- AspNetHostedService
                | -- Solution file 
                | -- README.md (currently sample_AspNetCore.md)
                | -- Directory.Build.targets  (version overrides when outside the repository; optional - doing it inline in each project is fine)
                | -- src
                        | -- sample project and assets, less the markdown file
                | -- tests
                       | -- test project (new)
                       | -- HostedServiceTests.cs (hoisted from main tests project)
```

I leave it to your discretion whether you'd like to include an explicit gitignore, editorconfig or any of the other "I downloaded this project in a zipfile" assets that the KeyVault sample has.  Likewise, I'm good if you'd prefer to keep the conditional version assignment locally in each project file or extract to a `Directory.Build.targets` file.

davetrainer replied:
Done. Left out gitignore and editorconfig

jsquire reviewed (commented):
Hi @davetrainer.  Thank you very much for your contribution and your interest in improving the Azure developer experience.   The sample looks great!   

I've left some feedback asking to align a few things closer to repository conventions, if you'd be so kind.

Commit:
pr feedback changes

Commit:
update to net8.0

Commit:
Merge branch 'main' into sample-aspnet-eventprocessorclient

Commit:
back to net7 due to sdk issues

Commit:
fix net462 build fail

davetrainer commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/Azure.Messaging.EventHubs.Samples.Processor.HostedService.csproj:
```diff
@@ -0,0 +1,15 @@
+    <ImplicitUsings>enable</ImplicitUsings>
+  </PropertyGroup>
+
+  <ItemGroup>
```
Directory.Build.targets added

davetrainer commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/Azure.Messaging.EventHubs.Processor.Tests.csproj:
```diff
@@ -25,6 +25,9 @@
     <ProjectReference Include="$(AzureCoreTestFramework)" />
     <ProjectReference Include="..\src\Azure.Messaging.EventHubs.Processor.csproj" />
   </ItemGroup>
+  <ItemGroup Condition="$(TargetFramework) != 'net462'">
```
Done

Commit:
update README

davetrainer commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/SampleProgram.cs:
```diff
@@ -0,0 +1,58 @@
+namespace Azure.Messaging.EventHubs.Samples.Processor.HostedService
+{
+    /// <summary>
+    /// Test
```
Done

davetrainer commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/SampleProgram.cs:
```diff
@@ -0,0 +1,58 @@
+
+            //Adds the hosted service to the service collection.
+            builder.Services.AddHostedService<EventProcessorClientService>();
+
```
Done

davetrainer commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/sample_AspNetCore.md:
````diff
@@ -0,0 +1,102 @@
+#### From the dotnet CLI
+
+1. Open a terminal in the root of the HostedService sample project folder.
+2. Run the following dotnet command: ```dotnet run --framework net7.0```
````
Left as is due to net8 SDK issue

davetrainer commented:
@jsquire Thanks for the detailed and thorough review! I've made the necessary changes. I was unable to get the solution building when targeting net8.0, due to an SDK mismatch. I wasn't bale to resolve it. If there's a way around this, I can incorporate it if needed.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/README.md:
```diff
@@ -81,6 +81,9 @@ For the event processor client to make use of Azure Storage blobs for checkpoint
 - [Mocking Client Types](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/Sample08_MockingClientTypes.md)  
   A demonstration of how to mock the types in the Event Hubs Processor client library, focusing on common application scenarios.
 
+- [Using the `EventProcessorClient` with an ASP.NET hosted service](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/README.md)  
```
This URL will likely continue to cause analysis to fail because it is correctly an explicit reference to `main`  but the path doesn't yet exist.   Once we get to the point where it's our only remaining failure, we can decide whether to force bypass of the failure or comment it out and bring it back after this merges.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/AspNetHostedService.sln:
Not sure if the path was already updated and not showing in the GH UI, but can we please confirm that we've adjusted case to `AspNetHostedService`?

davetrainer replied:
@jsquire Folder path is now correctly cased

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/Directory.Build.targets:
```diff
@@ -0,0 +1,9 @@
+  </ItemGroup>
+
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory).., Directory.Build.targets))\Directory.Build.targets"
+          Condition="'$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory).., Directory.Build.targets))' != ''" />
```
We don't want the condition here; this will need to link back up to the root targets file in order to keep the needed integration with the repo engineering process and systems.   We'll need to adjust the path to climb up a few more subdirectories.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/Directory.Build.targets:
```diff
@@ -0,0 +1,9 @@
+<Project>
+  <!-- Use decentralized package references when building outside https://github.com/Azure/azure-sdk-for-net -->
+  <ItemGroup Condition="'$(IsSamplesProject)' != 'true'">
```
We'll want to be sure to do this after the import, so that we're covered in case the definition of `IsSamplesProject` moves around to the repo targets instead of the props where it lives today.   Unlikely that it would happen, but there's no downside to flipping the order here.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/README.md:
```diff
@@ -0,0 +1,108 @@
+description: Using the EventProcessorClient with an ASP.NET hosted service
+---
+
+# Using the `EventProcessorClient` with an ASP.NET hosted service
```
```suggestion
# Using the `EventProcessorClient` with an ASP.NET hosted service

```

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/README.md:
```diff
@@ -0,0 +1,108 @@
+
+## Getting Started
+
+### Prerequisites
```
```suggestion
### Prerequisites

```

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/README.md:
```diff
@@ -0,0 +1,108 @@
+---
+
+# Using the `EventProcessorClient` with an ASP.NET hosted service
+This is an example of how to create and configure the `EventProcessorClient` in an ASP.NET context as an `IHostedService` implementation.
+ASP.NET hosted services can be used to implement background task processing, making them a good choice if your application needs to continually process events from Azure Event Hubs.  For more information, please see the [hosted services documentation](https://learn.microsoft.com/aspnet/core/fundamentals/host/hosted-services).
```
```suggestion
This is an example of how to create and configure the `EventProcessorClient` in an ASP.NET context as an `IHostedService` implementation.  ASP.NET hosted services can be used to implement background task processing, making them a good choice if your application needs to continually process events from Azure Event Hubs.  For more information, please see the [hosted services documentation](https://learn.microsoft.com/aspnet/core/fundamentals/host/hosted-services).
```

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/tests/Azure.Messaging.EventHubs.Processor.Samples.HostedService.Tests.csproj:
```diff
@@ -0,0 +1,24 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>$(RequiredTargetFrameworks)</TargetFrameworks>
```
This will fail outside the Azure SDK repository.   We'll need to do a conditional definition to define these if we're not in the repo.   Since we've got a targets file, that'd be the place.

davetrainer replied:
@jsquire I'm not fully understanding the ask here. Can you give me an example of what you're looking for? I've looked for other examples in sample projects but can't find anything.

davetrainer replied:
@jsquire Understood. I've added `Directory.Build.props` and updated `Directory.Build.targets`. I'm forced to only target net6 and net7 in `Directory.Build.props`, as it's not possible to build against net462 for the aspnet app (net462 get's passed down by `Directory.Build.Common.props` , but I have to override ride it).  If there's a way to work around this, please let me know.

jsquire replied:
The only way to work around it would be to screen it out unless `IsSamplesProject` is set;  the `net462` target is part of our required test matrix in the repo.   I think the approach that you've got for just hiding the tests is the best we can do.

jsquire reviewed (commented):
Thanks, @davetrainer.   We're looking good!  There's a few minor nits and some adjustments  for accomodating the dual nature of being inside and outside of the repo that I've asked for.

Commit:
fix issues with repo structure, folder rename, update README

Commit:
Merge branch 'main' into sample-aspnet-eventprocessorclient

davetrainer commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/src/SampleProgram.cs:
```diff
@@ -0,0 +1,69 @@
+        /// <summary>
+        /// The entry point to the application.
+        /// </summary>
+        /// <param name="args"></param>
```
@jsquire The SDK build system is enforcing XML comments, so I have to add something. I've updated where needed.

Commit:
Reintroduce net462 for build scenario and add IsPackable

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/Directory.Build.props:
```diff
@@ -0,0 +1,10 @@
+﻿<Project ToolsVersion="Current" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <!-- Import the common SDK build properties. -->
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory)..\..\..\.., Directory.Build.props))\Directory.Build.props"
```
```suggestion
  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory)..\..\.., Directory.Build.props))\Directory.Build.props"
```

If I'm reading correctly, I think this climbs 4 directories up, which puts us in the `/sdk` directory.  Since we want to be in `/sdk/eventhub`  I believe we need to trim one level.

Please double check me.

davetrainer replied:
@jsquire  I can confirmed this is the correct change.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/Directory.Build.props:
```diff
@@ -0,0 +1,10 @@
+﻿<Project ToolsVersion="Current" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <!-- Import the common SDK build properties. -->
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory)..\..\..\.., Directory.Build.props))\Directory.Build.props"
+			Condition="'$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory)..\..\..\.., Directory.Build.props))' != ''" />
```
```suggestion
			Condition="'$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory)..\..\.., Directory.Build.props))' != ''" />
```

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/tests/Azure.Messaging.EventHubs.Processor.HostedService.Samples.Tests.csproj:
```diff
@@ -0,0 +1,16 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFrameworks>$(RequiredTargetFrameworks);net462</TargetFrameworks>
```
Should we consider only including `net462` when building from the Azure SDK repo?  It really exists just to service our common text matrix.

davetrainer replied:
@jsquire Agreed. I've added this to the props file.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/Azure.Messaging.EventHubs.Processor.HostedService.Samples.csproj:
```diff
@@ -0,0 +1,15 @@
+﻿<Project Sdk="Microsoft.NET.Sdk.Web">
+
+  <PropertyGroup>
```
Can we please rename to `Azure.Messaging.EventHubs.Processor.Samples.HostedService`?   

The Event Hubs overrides allow for this pattern to also set the `IsSamplesProject` flag that we're using for detecting the build location.  _([src](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/eventhub/Directory.Build.props#L10))_

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/EventProcessorClientService.cs:
```diff
@@ -0,0 +1,178 @@
+        /// Implementation of the <see cref="IHostedService"/> StartAsync interface method.
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
```
```suggestion
```

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/EventProcessorClientService.cs:
```diff
@@ -0,0 +1,178 @@
+        /// <summary>
+        /// Implementation of the <see cref="IHostedService"/> StartAsync interface method.
+        /// </summary>
+        /// <param name="cancellationToken"></param>
```
Please add a description here.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/EventProcessorClientService.cs:
```diff
@@ -0,0 +1,178 @@
+            }
+            catch
+            {
+                // It is important that you guard against exceptions in
```
This is true of the event handlers assigned to the processor, but I think we want to throw here if the processor has issues starting, no?   For example, if the connection string is incorrect, we know that the processor will never recover.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/EventProcessorClientService.cs:
```diff
@@ -0,0 +1,178 @@
+        /// Implementation of the <see cref="IHostedService"/> StopAsync interface method.
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
```
```suggestion
```

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/EventProcessorClientService.cs:
```diff
@@ -0,0 +1,178 @@
+        /// <summary>
+        /// Implementation of the <see cref="IHostedService"/> StopAsync interface method.
+        /// </summary>
+        /// <param name="cancellationToken"></param>
```
Please add text for this one.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/SampleApplicationProcessor.cs:
```diff
@@ -0,0 +1,46 @@
+        public async Task Process(string body)
+        {
+            _logger.LogInformation("Event body has been processed: {body}", body);
+            await Task.CompletedTask.ConfigureAwait(false);
```
```suggestion
            return Task.CompletedTask;
```

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/SampleApplicationProcessor.cs:
```diff
@@ -0,0 +1,46 @@
+        /// Processes the string representation of an event. The sample behaviour logs the body using a named <see cref="ILogger"/>.
+        /// </summary>
+        /// <param name="body">The body of an event to process.</param>
+        /// <returns></returns>
```
```suggestion
```

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/SampleApplicationProcessor.cs:
```diff
@@ -0,0 +1,46 @@
+        /// The default method of the <see cref="ISampleApplicationProcessor"/> interface used for processing of event data. Implement this method in a class definition.
+        /// </summary>
+        /// <param name="body">The body of an event to process.</param>
+        /// <returns></returns>
```
```suggestion
```

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/SampleProgram.cs:
```diff
@@ -0,0 +1,69 @@
+        /// <summary>
+        /// The default entry point method of the <see cref="SampleProgram" /> class.
+        /// </summary>
+        /// <param name="args"></param>
```
We may want to remove the args parameter, since we're not actually accepting anything.   Otherwise, please add text to the doc comment.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/SampleProgram.cs:
```diff
@@ -0,0 +1,69 @@
+            app.Run();
+        }
+
+        private static EventProcessorClient GetEventProcessorClient(IConfiguration configuration)
```
```suggestion
        private static EventProcessorClient CreateEventProcessorClient(IConfiguration configuration)
```
nit: "Get" could be interpreted as grabbing an existing instance from the DI container or other cache.

jsquire reviewed (commented):
Apologies for the delay, @davetrainer.   This is looking great!   I started digging into the lower-level details this time out, and there are just a few small nits that I left feedback on.   I think we're just about ready!

Commit:
Merge branch 'main' into sample-aspnet-eventprocessorclient

Commit:
rename folder

Commit:
rename projects and update props path

Commit:
address code comments

davetrainer commented:
> Apologies for the delay, @davetrainer. This is looking great! I started digging into the lower-level details this time out, and there are just a few small nits that I left feedback on. I think we're just about ready!

Hi @jsquire Happy new year! I have addressed all comments. Could you please take another look? It would be great if I could get this merged.

jsquire commented on eng/Packages.Data.props:
```diff
@@ -288,9 +289,11 @@
     <PackageReference Update="Microsoft.Azure.WebJobs.Extensions.Http" Version="3.0.2" />
     <PackageReference Update="Microsoft.CSharp" Version="4.7.0" />
     <PackageReference Update="Microsoft.Extensions.Azure" Version="1.6.3" />
+	<PackageReference Update="Microsoft.Extensions.Configuration.Abstractions" Version="7.0.0" />
```
Any reason that we shouldn't push these to the latest?   I believe the 8.0 line should work just fine with `net6.0` and `net7.0` targets.  Not particularly important nor blocking.   I can always bump them in a follow-up.

davetrainer replied:
@jsquire 7.0 would've been what I got when VS added the nuget package reference - I assume as it's targets net 7, it picked that one. I'm happy to bump and retest if you want me to?

jsquire replied:
@davetrainer:  I appreciate your willingness, but no need.   I'll need to do a follow-up to uncomment the new sample link once it's live.   I'll play with the package references when I do that one.   I'm just waiting on one other thing to close internally and then I'll get this merged in.

jsquire commented:
> Hi @jsquire Happy new year! I have addressed all comments. Could you please take another look? It would be great if I could get this merged.
> 
> I've tested inside and out of the repo structure, the app works as expected in both.

Thanks, @davetrainer!   Your efforts and contribution are very much appreciated.   This looks great.   I'm kicking the (unrelated) failed bits in CI to see if I can get them to pass on re-run.   If not, I'll just override the failure and get things merged!

jsquire commented on eng/Packages.Data.props:
```diff
@@ -288,9 +289,11 @@
     <PackageReference Update="Microsoft.Azure.WebJobs.Extensions.Http" Version="3.0.2" />
     <PackageReference Update="Microsoft.CSharp" Version="4.7.0" />
     <PackageReference Update="Microsoft.Extensions.Azure" Version="1.6.3" />
+	<PackageReference Update="Microsoft.Extensions.Configuration.Abstractions" Version="7.0.0" />
```
```suggestion
	 <PackageReference Update="Microsoft.Extensions.Configuration.Abstractions" Version="8.0.0" />
```

jsquire commented on eng/Packages.Data.props:
```diff
@@ -288,9 +289,11 @@
     <PackageReference Update="Microsoft.Extensions.Configuration.Binder" Version="2.1.10" />
     <PackageReference Update="Microsoft.Extensions.Configuration.Json" Version="5.0.0" />
     <PackageReference Update="Microsoft.Extensions.Configuration" Version="5.0.0" />
+    <PackageReference Update="Microsoft.Extensions.Hosting" Version="7.0.1" />	
```
```suggestion
    <PackageReference Update="Microsoft.Extensions.Hosting" Version="8.0.0" />	
```

jsquire committed:
Update eng/Packages.Data.props

jsquire committed:
Update eng/Packages.Data.props
